pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -214,40 +234,239 @@ func (c *MCRouteController) Run(stopCh <-chan struct{}) {
+		}
+	}
+
+	if err := c.syncMCFlows(); err == nil {
+		c.ciImportQueue.Forget(obj)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.ciImportQueue.AddRateLimited(ciImp)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", ciImp)
 	}
+	return true
 }
 
-func (c *MCRouteController) processNextWorkItem() bool {
-	obj, quit := c.queue.Get()
+func (c *MCDefaultRouteController) processNextGatewayItem() bool {
+	obj, quit := c.gatewayQueue.Get()
 	if quit {
 		return false
 	}
-	defer c.queue.Done(obj)
+	defer c.gatewayQueue.Done(obj)
 
-	if k, ok := obj.(string); !ok {
-		c.queue.Forget(obj)
+	gateway, ok := obj.(string)
+	if !ok {
+		c.gatewayQueue.Forget(obj)
 		klog.InfoS("Expected string in work queue but got", "object", obj)
 		return true
-	} else if err := c.syncMCFlows(); err == nil {
-		c.queue.Forget(k)
+	}
+
+	if c.wireGuardConfig != nil {
+		if err := c.reconcileWireGuard(); err != nil {
+			klog.ErrorS(err, "Error when reconcile WireGuard, requeuing", "key", gateway)
+			c.gatewayQueue.AddRateLimited(gateway)
+		}
+	}
+	if err := c.syncMCFlows(); err == nil {
+		c.gatewayQueue.Forget(obj)
 	} else {
 		// Put the item back on the workqueue to handle any transient errors.
-		c.queue.AddRateLimited(k)
-		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+		c.gatewayQueue.AddRateLimited(gateway)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", gateway)
 	}
 	return true
 }
 
-func (c *MCRouteController) syncMCFlows() error {
+func (c *MCDefaultRouteController) reconcileWireGuard() error {
+	c.wireGuardMutex.Lock()
+	defer c.wireGuardMutex.Unlock()
+	gateways, err := c.gwLister.Gateways(c.namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	if c.wireGuardInitialized {
+		// The Gateway is deleted. Remove WireGuard interface
+		if len(gateways) == 0 || gateways[0].Name != c.nodeConfig.Name {
+			klog.InfoS("Gateway is deleted, removing WireGuard interface and rule")
+			if err := c.deleteWireGuardInterfaceAndRule(); err != nil {
+				return err
+			}
+			c.wireGuardInitialized = false
+			return nil
+		}
+	}
+
+	// If the WireGuard is not initialized, and the Node is the Gateway Node.
+	// Controller will create corresponding interface and route on the host.
+	if !c.wireGuardInitialized && len(gateways) != 0 && gateways[0].Name == c.nodeConfig.Name {
+		klog.InfoS("Initializing WireGuard interface and rule")
+		if err := c.initializeWireGuard(); err != nil {
+			return err
+		}
+		if err := c.initWireGuardRoute(); err != nil {
+			return err
+		}
+		c.wireGuardInitialized = true
+
+		// Add existing ciImports to WireGuard Peers
+		return c.updateWireGuardPeers("")
+	}
+
+	return nil
+}
+
+func (c *MCDefaultRouteController) deleteWireGuardInterfaceAndRule() error {
+	if err := c.routeClient.DeleteIPRule(multiclusterWireGuardFwmark, multiclusterWireGuardRouteTable); err != nil {
+		return err
+	}
+	return c.wireGuardClient.Cleanup()
```

Comment:
wireGuardClient.Cleanup() just deletes the interface. Do we need to clean up for the added peers too?

Guess we should set c.wireGuardClient to nil too.

Reply:
Set `c.wireGuardClient` to nil now.

Reply:
Add a cleanup defer function now. When init, if the process fails partly, it will delete resources(eg. interface).

Reply:
Cleanup operations wont fail?